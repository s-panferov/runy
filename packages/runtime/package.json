{
  "name": "@runy-build/runtime",
  "version": "0.1.0",
  "type": "module",
  "license": "MIT",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "dependencies": {
    "@runy-build/schema": "workspace:*",
    "execa": "^9.5.3",
    "vscode-jsonrpc": "^8.2.1",
    "winston": "^3.17.0"
  },
  "exports": {
    ".": "./dist/index.js",
    "./BUILD": "./BUILD.ts"
  },
  "files": [
    "dist/index.js",
    "dist/index.js.map",
    "dist/index.d.ts",
    "package.json",
    "README.md",
    "LICENSE"
  ]
}
