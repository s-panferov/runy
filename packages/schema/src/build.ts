import { Compute, Provide, Result } from "./index";

import { BuildContext } from "./context";
import { Schema } from "./schema";
import { BuildSpec } from "./spec";
import { Target, TargetT } from "./target";
import { COMPUTE_SYM, PROVIDE_SYM, RESULT_SYM, TO_SCHEMA } from "./symbols";
import { dynamic, DYNAMIC } from "./missing";
import { Output, Query } from "./output";
import { hash } from "./hash";

export type BuildPhase = "out" | "ready";

export class BuildRef<T extends TargetT, P extends BuildPhase>
  implements Compute
{
  [COMPUTE_SYM]!: true;
  [RESULT_SYM]!: Result<T["Out"]>;
  [PROVIDE_SYM]!: Provide<T["Out"]>;

  hash: string = "";

  private constructor(
    public ctx: BuildContext,
    public target: Target<T>,
    public spec: BuildSpec<T>,
    public phase: P
  ) {}

  static dyn<T extends TargetT, P extends BuildPhase>(
    ctx: BuildContext,
    target: Target<T>,
    spec: BuildSpec<T>,
    phase: P
  ): Query<P extends "out" ? T["Out"] : T["Ready"]> {
    return dynamic(new BuildRef<T, P>(ctx, target, spec, phase)) as Query<
      P extends "out" ? T["Out"] : T["Ready"]
    >;
  }

  computeSchemaObject(schema: Schema) {
    return {
      target: schema.convert(this.target),
      ctx: schema.convert(this.ctx),
      spec: schema.convert(this.spec),
    };
  }

  [TO_SCHEMA] = (schema: Schema): object => {
    let json: object | undefined;
    if (!this.hash) {
      json = this.computeSchemaObject(schema);
      this.hash = hash(json);
    }

    if (!schema.builds[this.hash]) {
      const obj = {} as any;
      schema.builds[this.hash] = obj;
      Object.assign(obj, json || this.computeSchemaObject(schema));
    }

    return { kind: "build", $ref: this.hash, phase: this.phase };
  };

  [DYNAMIC] = (prop: string | symbol) => {
    if (prop == TO_SCHEMA) {
      return this[TO_SCHEMA];
    }

    return Output.dyn(this, prop as string);
  };
}
