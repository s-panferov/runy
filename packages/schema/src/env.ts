import { dirname } from "node:path";
import process from "node:process";
import assert from "node:assert";

export let WORKSPACE_ROOT = "";

try {
  WORKSPACE_ROOT = process.env["BUILD_WORKSPACE"] || process.cwd();
} catch {
  // no access
}

export function setWorkspaceRoot(path: string, parent = false): void {
  assert(!!path);

  // Check if the path is a URL
  try {
    const url = new URL(path);
    // If it's a valid URL, convert it to a pathname and use its parent
    WORKSPACE_ROOT = url.pathname;
  } catch {
    // Use the parent path by default
    WORKSPACE_ROOT = path;
  }

  if (parent) {
    WORKSPACE_ROOT = dirname(WORKSPACE_ROOT);
  }
}

export function setTestRoot(path: string) {
  assert(!!path);
  setWorkspaceRoot(path, true);
}
