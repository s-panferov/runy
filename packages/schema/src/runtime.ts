import { Compute } from ".";
import { Query } from "./output";
import { Schema, TO_SCHEMA, ToSchema } from "./schema";
import { COMPUTE_SYM, PROVIDE_SYM, RESULT_SYM } from "./symbols";

type Var = Query<string | number | object>;

export interface RuntimeOptions {
  path: string;
}

export class Runtime implements ToSchema, Compute {
  [COMPUTE_SYM]!: true;
  [PROVIDE_SYM]!: string;
  [RESULT_SYM]!: string;

  private options: RuntimeOptions;

  constructor(options: RuntimeOptions) {
    this.options = options;
  }

  [TO_SCHEMA](_schema: Schema) {
    let value: any = {
      kind: "runtime",
      path: this.options.path,
    };

    return value;
  }
}
