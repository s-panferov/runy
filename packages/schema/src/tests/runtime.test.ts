import { BuildContext, Schema } from "../index";
import { setWorkspaceRoot } from "../env";
import { expect, test, describe } from "vitest";

import { target } from "./runtime.build";

describe("runtime input", () => {
  test("runtime input", () => {
    setWorkspaceRoot(import.meta.url);

    const schema = Schema.convert(target().out(new BuildContext()));
    console.log(schema);

    expect(schema).toMatchSnapshot();
  });
});
