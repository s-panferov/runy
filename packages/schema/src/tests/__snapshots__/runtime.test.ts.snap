// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runtime input > runtime input 1`] = `
{
  "builds": {
    "uVs0GtLQLI/lX18mDOMhYYVxEhA=": {
      "ctx": {
        "$ref": "U1wwAaqukwfarhkuFGe2QSaTdXY=",
        "kind": "context",
      },
      "spec": {
        "inout": {},
        "inp": {
          "files": [
            {
              "kind": "fileset",
              "path": [
                "src/**/*",
              ],
            },
          ],
        },
        "key": [],
        "out": {
          "folder": [
            {
              "kind": "folder",
              "path": undefined,
            },
          ],
          "out": {
            "kind": "runtime",
            "path": "./runtime.rt",
          },
        },
        "ready": {},
      },
      "target": {
        "$ref": "sq5PL1u1j6Ac02MnPoRLDMtzpnw=",
        "kind": "target",
      },
    },
  },
  "contexts": {
    "U1wwAaqukwfarhkuFGe2QSaTdXY=": {
      "flags": {},
      "parent": undefined,
    },
  },
  "flags": {},
  "root": {
    "$ref": "uVs0GtLQLI/lX18mDOMhYYVxEhA=",
    "kind": "build",
    "phase": "out",
  },
  "targets": {
    "sq5PL1u1j6Ac02MnPoRLDMtzpnw=": {
      "default": false,
      "name": "target",
      "package": "../runtime.build.ts",
    },
  },
}
`;
